agentendpoint/agentendpoint.go

```diff
@@ -0,0 +1,221 @@
+)
+
+var (
+	errServerCancel = errors.New("task canceled by server")
+	taskStateFile   = config.TaskStateFile()
+	patchEnabled    = config.OSPatchEnabled()
+)
+
+// Client is a an agentendpoint client.
+type Client struct {
+	raw *agentendpoint.Client
+}
+
+var client *Client
+
+// NewClient a new agentendpoint Client.
+func NewClient(ctx context.Context) (*Client, error) {
+	c, err := agentendpoint.NewClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	client = &Client{raw: c}
+	return client, nil
+}
+
+// ReportTaskProgress calls the agentendpoint service ReportTaskProgress call injecting the InstanceIdToken.
+func (c *Client) ReportTaskProgress(ctx context.Context, req *agentendpointpb.ReportTaskProgressRequest) (res *agentendpointpb.ReportTaskProgressResponse, err error) {
+	logger.Debugf("ReportTaskProgress request:\n%s", util.PrettyFmt(req))
+	token, err := config.IDToken()
+	if err != nil {
+		return nil, err
+	}
+	req.InstanceIdToken = token
+	if err := retryAPICall(2100*time.Second, "ReportTaskProgress", func() error {
+		res, err = c.raw.ReportTaskProgress(ctx, req)
+		if err != nil {
+			return err
+		}
+		logger.Debugf("ReportTaskProgress response:\n%s", util.PrettyFmt(res))
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error calling ReportTaskProgress: %v", err)
+	}
+
+	return
+}
+
+// ReportTaskComplete calls the agentendpoint service ReportTaskComplete call injecting the InstanceIdToken.
+func (c *Client) ReportTaskComplete(ctx context.Context, req *agentendpointpb.ReportTaskCompleteRequest) error {
+	logger.Debugf("ReportTaskComplete request:\n%s", util.PrettyFmt(req))
+	token, err := config.IDToken()
+	if err != nil {
+		return err
+	}
+	req.InstanceIdToken = token
+	if err := retryAPICall(2100*time.Second, "ReportTaskComplete", func() error {
+		res, err := c.raw.ReportTaskComplete(ctx, req)
+		if err != nil {
+			return err
+		}
+		logger.Debugf("ReportTaskComplete response:\n%s", util.PrettyFmt(res))
+		return nil
+	}); err != nil {
+		return fmt.Errorf("error calling ReportTaskComplete: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Client) runTask(ctx context.Context) {
+	for {
+		token, err := config.IDToken()
+		if err != nil {
+			logger.Errorf(err.Error())
+			return
+		}
+
+		start := &agentendpointpb.ReportTaskStartRequest{
+			InstanceIdToken: token,
+		}
+		startRes, err := c.raw.ReportTaskStart(ctx, start)
+		if err != nil {
+			logger.Errorf(err.Error())
+			return
+		}
+
+		task := startRes.GetTask()
+		if task == nil {
+			return
+		}
+
+		switch task.GetTaskType() {
+		case agentendpointpb.TaskType_APPLY_PATCHES:
+			if !patchEnabled {
+				logger.Infof("Recieved TaskType_APPLY_PATCHES but OSPatch is not enabled")
+				continue
+			}
+
+			if err := c.RunApplyPatches(ctx, task.GetApplyPatchesTask()); err != nil {
+				logger.Errorf(err.Error())
+			}
+		case agentendpointpb.TaskType_EXEC_STEP_TASK:
+			if err := c.RunExecStep(ctx, task.GetExecStepTask()); err != nil {
```

Comment:
If patch is not enabled, we shouldn't be allowing exec step as well, right?

Reply:
Actually this didn't make much sense, I removed the check from here and will add it in to the caller. If patch isn't enabled we don't even want the endpoint setup right now. Ill need to rethink this so just dropped from this PR